Show check results on admin page

jjgrainger opened the pull request:
Addresses acceptance criteria in #35 

Closes #35

jjgrainger committed:
create results table templates

jjgrainger committed:
add results to page

jjgrainger committed:
Merge branch 'trunk' into feature/ajax-show-check-results

spacedmonkey commented on templates/results-table.php:
```diff
@@ -0,0 +1,32 @@
+			</tr>
+			<tr>
+				<td>
+					Line
```
Should these be translated?

jjgrainger replied:
Yes! Thanks for flagging. I've updated these.

spacedmonkey commented on assets/js/plugin-check-admin.js:
```diff
@@ -207,4 +211,89 @@
+			'plugin-check__results-body-' + index
+		);
+
+		// Loop over each result by the line, column and messages.
+		for ( const line in errors ) {
+			for ( const column in errors[ line ] ) {
+				for ( let i = 0; i < errors[ line ][ column ].length; i++ ) {
+					const message = errors[ line ][ column ][ i ].message;
+					const code = errors[ line ][ column ][ i ].code;
+
+					resultsTable.innerHTML += rowTemplate( {
+						line,
+						column,
+						file,
+						type: 'ERROR',
+						message,
+						code,
+					} );
+				}
+			}
+		}
```
Repeated code again. There only two params that are different here are `errors` and the row type 'ERROR'. 

This should be very simple to change to a function.

jjgrainger replied:
Thanks @spacedmonkey I've created a function for this.

spacedmonkey commented:
@jjgrainger There is no styling on the table. Is this expected?

![Screenshot 2023-03-31 at 14 48 59](https://user-images.githubusercontent.com/237508/229138397-9c932b16-4d66-4f6d-84ef-75d5d52fca48.png)

spacedmonkey commented on templates/results-table.php:
```diff
@@ -0,0 +1,32 @@
+<script type="text/template" id="tmpl-plugin-check-results-table">
+	<table id="plugin-check__results-table-{{data.index}}" class="plugin-check__results-table">
```
```suggestion
	<table id="plugin-check__results-table-{{data.index}}" class="widefat plugin-check__results-table">
```

jjgrainger replied:
Thanks @spacedmonkey updated this with your suggestion

spacedmonkey commented on templates/results-table.php:
```diff
@@ -0,0 +1,32 @@
+<script type="text/template" id="tmpl-plugin-check-results-table">
+	<table id="plugin-check__results-table-{{data.index}}" class="plugin-check__results-table">
+		<thead>
+			<tr>
+				<td>
+					FILE:
+				</td>
+				<td colspan="4">
+					{{ data.file }}
+				</td>
+			</tr>
```
I am not sure that this makes sense as a table header. Maybe H tag makes more tense here.

jjgrainger replied:
Thanks @spacedmonkey updated to a H tag and this looks better with the table styling too.

spacedmonkey commented:
![Screenshot 2023-03-31 at 14 53 06](https://user-images.githubusercontent.com/237508/229139246-cdd80ca0-f4d1-4d6c-8232-0fdeca5bfeb8.png)

When I click on a plugin without an error or warning, I see no message at all. It seems like the plugin checker has done nothing here. This maybe hard to test, if plugin even works.

spacedmonkey commented on includes/Admin/Admin_Page.php:
```diff
@@ -154,4 +157,14 @@ public function filter_plugin_action_links( $actions, $plugin_file ) {
+	 * @since n.e.x.t
+	 */
+	public function admin_footer() {
+		include_once WP_PLUGIN_CHECK_PLUGIN_DIR_PATH . '/templates/results-table.php';
+		include_once WP_PLUGIN_CHECK_PLUGIN_DIR_PATH . '/templates/results-row.php';
```
Why include over require?

jjgrainger replied:
Updated to require

spacedmonkey commented on templates/results-table.php:
```diff
@@ -0,0 +1,32 @@
+			</tr>
+		</thead>
+		<tbody id="plugin-check__results-body-{{data.index}}"></tbody>
+	</table>
```
Maybe we could add a <br /> or other spacer after each table.

jjgrainger replied:
Thanks @spacedmonkey I've added a line break after each table. This along with using a heading for the filename gives it spacing to make it easier to read.

jjgrainger committed:
update table layout and style

jjgrainger committed:
create renderResultRows function

jjgrainger committed:
add message once plugin check process is complete

jjgrainger commented:
Thanks @spacedmonkey 

I've updated the PR based on your feedback. As for your comment...

> When I click on a plugin without an error or warning, I see no message at all. It seems like the plugin checker has done nothing here. This maybe hard to test, if plugin even works.

I agree, and for now I've added a simple 'Checks complete' message once the process is done. I think it would be good to discuss what output through the process we'd want to display as part of the milestone review. Can you create an issue for this one?

spacedmonkey commented:
![Screenshot 2023-03-31 at 18 10 39](https://user-images.githubusercontent.com/237508/229185930-232c4999-f25e-43c0-a43f-2bfad7da4f63.png)

This is looking a lot better! 

Couple of things. If I run this a couple of times, I see complete message twice, I don't know why this is. 

Another nitpic, but would not take long to implement, when you click the button, if you a spinner could show, then go away when request is complete that would be amazing. Jetpack took a while to process and did not know when it was complete. To add a spinner in WP is very simple. 

Next to the submit button add this. 

```
<span class="spinner"></span>
```

Add a id to it. When the button is clicked, add `is-active` class to the id and remove when the requests are complete. Should not take more than a couple of minutes to add and will make it clear to the user that something is happening.

felixarntz commented on assets/js/plugin-check-admin.js:
```diff
@@ -207,4 +213,85 @@
+	 * @param {Object} warnings The file warnings.
+	 */
+	function renderFileResults( file, errors, warnings ) {
+		const tableTemplate = wp.template( 'plugin-check-results-table' );
```
Parsing the template here will result in the same template being parsed several times, we should avoid that.

spacedmonkey commented on templates/results-table.php:
```diff
@@ -0,0 +1,26 @@
+<script type="text/template" id="tmpl-plugin-check-results-table">
+	<h4><?php esc_html_e( 'FILE:', 'plugin-check' ); ?> {{ data.file }}</h4>
+	<table id="plugin-check__results-table-{{data.index}}" class="widefat plugin-check__results-table">
```
```suggestion
	<table id="plugin-check__results-table-{{data.index}}" class="widefat striped plugin-check__results-table">
```

felixarntz commented on assets/js/plugin-check-admin.js:
```diff
@@ -207,4 +213,85 @@
+	 * @param {Object} table   The HTML table to append a result row to.
+	 */
+	function renderResultRows( type, results, table ) {
+		const rowTemplate = wp.template( 'plugin-check-results-row' );
```
Same here, since this function is also called several times.

I think to fix that in both instance, we should introduce a general function that handles the template rendering generically and is smart about it, something like:
```js
const templates = {};
function renderTemplate( templateSlug, data ) {
	if ( ! templates[ templateSlug ] ) {
		templates[ templateSlug ] = wp.template( templateSlug );
	}
	const template = templates[ templateSlug ];
	return template( data );
}
```

spacedmonkey committed:
Apply suggestions from code review

felixarntz commented on assets/js/plugin-check-admin.js:
```diff
@@ -1,24 +1,30 @@
 			.then( ( data ) => {
 				console.log( data.message );
+
+				resultsContainer.innerHTML += 'Checks complete';
```
This should be a translated string, also we should probably put it at least in some minimal markup, like a `<p>` tag.

Last but not least, having this line here seems out of context. Shouldn't it rather happen as the last bit of `renderResults()` as it is connected to that function?

felixarntz commented on assets/js/plugin-check-admin.js:
```diff
@@ -1,24 +1,30 @@
 		e.preventDefault();
 
+		// Empty the results container.
+		resultsContainer.innerText = '';
```
Let's move this into a simple function like `resetResults()`.

spacedmonkey reviewed (dismissed):
I think barring the spinner, this is nearly there.

felixarntz commented on includes/Admin/Admin_Page.php:
```diff
@@ -154,4 +157,14 @@ public function filter_plugin_action_links( $actions, $plugin_file ) {
+	 * @since n.e.x.t
+	 */
+	public function admin_footer() {
+		require_once WP_PLUGIN_CHECK_PLUGIN_DIR_PATH . '/templates/results-table.php';
+		require_once WP_PLUGIN_CHECK_PLUGIN_DIR_PATH . '/templates/results-row.php';
```
A few things here:
* These files should preferably only include the templates themselves, not the surrounding script tag. For that we should use the `wp_print_inline_script_tag()` function. To pass the output of these templates into the function, we can then use an output buffer.
* No need for `require_once`, this file doesn't contain PHP functions or anything like that and therefore is okay to be included multiple times.
```suggestion
		ob_start();
		require WP_PLUGIN_CHECK_PLUGIN_DIR_PATH . '/templates/results-table.php';
		$results_table_template = ob_get_clean();
		wp_print_inline_script_tag(
			$results_table_template,
			array(
				'id'   => 'tmpl-plugin-check-results-table',
				'type' => 'text/template',
			)
		);

		ob_start();
		require WP_PLUGIN_CHECK_PLUGIN_DIR_PATH . '/templates/results-row.php';
		$results_row_template = ob_get_clean();
		wp_print_inline_script_tag(
			$results_row_template,
			array(
				'id'   => 'tmpl-plugin-check-results-row',
				'type' => 'text/template',
			)
		);
```

felixarntz reviewed (commented):
@jjgrainger Left some feedback here and above, but most of it looks good so far.

jjgrainger committed:
add spinner

jjgrainger committed:
add renderTemplate function

jjgrainger committed:
add resetResults function

jjgrainger committed:
update results templates

felixarntz commented on assets/js/plugin-check-admin.js:
```diff
@@ -207,4 +222,108 @@
+		}
+
+		resultsContainer.innerHTML +=
+			'<p>' + wp.i18n.__( 'Checks complete', 'plugin-check' ) + '</p>';
```
Providing translations to JS is a bit trickier, and since this is the only translated string we have in JS, I would prefer if we handled that in PHP and made it available to JS.

While that may seem a bit overkill, why don't we put this `<p>` tag with the string in another `script[type="text/template"]` and then render it like that? That would then fully decouple the UI from the JS logic.

felixarntz reviewed (commented):
@jjgrainger One last comment here, but this is pretty much good to go.

jjgrainger committed:
move complete message to template

joemcgill reviewed (commented):
This is looking good to me. I may have found a small bug that we could improve. It looks like if you click the button more than once wild checks are running, you will end up with multiple instances of the results. Probably the best way to handle this is to disable the button while the spinner is active. 

<img width="675" alt="image" src="https://user-images.githubusercontent.com/801097/229573009-6fe8902c-6900-44e5-b77b-2268257b2f5c.png">

felixarntz reviewed (approved):
@jjgrainger Thank you for the updates. Approved, pending the feedback from @joemcgill, it would be great to make the button temporarily `disabled` until the process is complete (or has terminated due to an error).

jjgrainger committed:
disable submit button on click

joemcgill reviewed (approved):
The latest commit fixes the double results issue for me. Thanks, @jjgrainger!
